crates/storage/provider/src/providers/state/historical.rs

```diff
@@ -31,46 +31,93 @@ pub struct HistoricalStateProviderRef<'a, 'b, TX: DbTx<'a>> {
     _phantom: PhantomData<&'a TX>,
 }
 
+pub enum HistoryInfo {
+    NotWritten,
+    InChangeset(u64),
+    InPlainState,
+}
+
 impl<'a, 'b, TX: DbTx<'a>> HistoricalStateProviderRef<'a, 'b, TX> {
     /// Create new StateProvider from history transaction number
     pub fn new(tx: &'b TX, block_number: BlockNumber) -> Self {
         Self { tx, block_number, _phantom: PhantomData {} }
     }
-}
-impl<'a, 'b, TX: DbTx<'a>> AccountProvider for HistoricalStateProviderRef<'a, 'b, TX> {
-    /// Get basic account information.
-    fn basic_account(&self, address: Address) -> Result<Option<Account>> {
+
+    /// Lookup an account in the AccountHistory table
+    pub fn account_history_lookup(&self, address: Address) -> Result<HistoryInfo> {
         // history key to search IntegerList of block number changesets.
         let history_key = ShardedKey::new(address, self.block_number);
+        let mut cursor = self.tx.cursor_read::<tables::AccountHistory>()?;
+
+        if let Some(chunk) =
+            cursor.seek(history_key)?.filter(|(key, _)| key.key == address).map(|x| x.1 .0)
+        {
+            let chunk = chunk.enable_rank();
+            let rank = chunk.rank(self.block_number as usize);
+            if rank == 0 && !cursor.prev()?.is_some_and(|(key, _)| key.key == address) {
+                return Ok(HistoryInfo::NotWritten)
+            }
+            if rank < chunk.len() {
+                Ok(HistoryInfo::InChangeset(chunk.select(rank) as u64))
+            } else {
+                Ok(HistoryInfo::InPlainState)
+            }
+        } else {
+            Ok(HistoryInfo::NotWritten)
+        }
```

Comment:
It could be helpful if we had some inline ascii diagram showing how this works, I think basically it says: 

If there's a value in the history table then it is:
* Not Written: If there are 0 elements in that entry AND there exists a previous entry in the table for that key, meaning it was previously written and now there was no change. We use this later for the optimization in the changeset read. In practice this is going to be rare as in most cases it will be`rank != 0` and the if statement will short circuit.
* InChangeset: If the number of items less than block number in that bucket is less than the items in the  bucket, return the corresponding changeset key.
* InPlainState: If it's larger than the bucket len, then it means it's in the plain state

If there's no value, also return NotWritten.